File: leetcode/2661-first-completely-painted-row-or-column/solution.cpp
/**
 * Author: 1chooo<hugo970217@gmail.com>
 * Problem: https://leetcode.com/problems/first-completely-painted-row-or-column
 * Runtime: 90ms (73.78%)
 */

const static auto _ = []() {
    cin.tie(nullptr)->sync_with_stdio(false);
    return nullptr;
}();

class Solution {
  public:
    int firstCompleteIndex(
        vector<int> &arr, vector<vector<int>> &mat) {
        int row = mat.size();
        int col = mat[0].size();
        unordered_map<int, pair<int, int>> positionMap;
        vector<int> rowCount(row, col);
        vector<int> colCount(col, row);

        for (int r = 0; r < row; ++r) {
            for (int c = 0; c < col; ++c) {
                positionMap[mat[r][c]] = {r, c};
            }
        }

        for (int i = 0; i < arr.size(); ++i) {
            int val = arr[i];
            auto [row, col] = positionMap[val];

            if (--rowCount[row] == 0 || --colCount[col] == 0) {
                return i;
            }
        }

        return -1;
    }
};

File: leetcode/2661-first-completely-painted-row-or-column/solution.py
"""
* Author: 1chooo<hugo970217@gmail.com>
* Problem: https://leetcode.com/problems/first-completely-painted-row-or-column
* Runtime: 119ms (67.19%)
"""

class Solution:
    def firstCompleteIndex(self, arr: List[int], mat: List[List[int]]) -> int:
        row: int = len(mat)
        col: int = len(mat[0])
        position_map: Dict[int, Tuple[int, int]] = {}
        for i in range(row):
            for j in range(col):
                position_map[mat[i][j]] = (i, j)

        row_count: List[int] = [col] * row
        col_count: List[int] = [row] * col

        for i, val in enumerate(arr):
            row, col = position_map[val]
            row_count[row] -= 1
            col_count[col] -= 1
            if row_count[row] == 0 or col_count[col] == 0:
                return i
            
        return -1


class Solution:
    def firstCompleteIndex(self, arr: List[int], mat: List[List[int]]) -> int:
        rows, cols = len(mat), len(mat[0])
        position_map = {mat[r][c]: (r, c) for r in range(rows) for c in range(cols)}
        row_count = [cols] * rows
        col_count = [rows] * cols
        for idx, val in enumerate(arr):
            row, col = position_map[val]
            row_count[row] -= 1
            col_count[col] -= 1
            if row_count[row] == 0 or col_count[col] == 0:
                return idx
        return -1